### PR TITLE
Add the text-file flag

### PR DESCRIPTION
**Pull request by mvdnes**
This adds the text flag which is stored in the internal file attributes.

### CODE REVIEW

**Comment by mvdnes** [comment]
This is in response to the difference in the zipinfo output as seen in #151. This does not solve that bug however.

File: src/write.rs [hunk 1]
```diff
@@ -133,6 +134,13 @@ impl<W: Write + io::Seek> Write for ZipWriter<W> {
                 "No file has been started",
             ));
         }
+
+        // Check if the buffer is text (UTF8) or contains binary data. Note that if an UTF8
+        // character is split between to calls to write, this will falsly mark it as binary.
+        // For ASCII characters, this is always correct.
```
Comment (rylev) [inline src/write.rs:140]: Can we use `Utf8Error::valid_up_to` to figure out if we’re in this situation? I think to know for sure we would have to keep the bytes from any previous read the are potentially a non-complete character and if validation fails, prepend and check again.

File: src/types.rs [hunk 2]
```diff
@@ -231,6 +231,8 @@ pub struct ZipFileData {
     pub header_start: u64,
     /// Specifies where the compressed data of the file starts
     pub data_start: u64,
+    /// Internal file attributes
+    pub internal_attributes: u16,
```
Comment (rylev) [inline src/types.rs:235]: This is out of scope for this PR but I wonder if keeping this internally as a bitflags enum would be better....

File: src/write.rs [hunk 3]
```diff
@@ -133,6 +134,13 @@ impl<W: Write + io::Seek> Write for ZipWriter<W> {
...
+        if self.stats.is_text == true && std::str::from_utf8(buf).is_err() {
+            self.stats.is_text = false;
+        }
```
Comment (Plecra) [inline src/write.rs:143]: This behavior seems annoying - if the user has indicated that they're *trying* to create a text file, they probably don't want it to just silently fail. 

I'd like to investigate using `fmt::Write` to require the user to validate the input (`from_utf8_unchecked` when they want to be quick), but maybe there's a better solution.

For now, can we just return an error here instead? *something* is wrong with the input we've been given, and the user should know

**Comment by Plecra** [comment]
I'd rather not provide this API like this - the issue came up because there are cases where it is significant whether or not this bit is set. There are cases where it will also be important that the flag *isn't* set, and this only leaves the user the option to deliberately provide invalid UTF-8 to get the flag set

File: src/write.rs (hunk 1 above)
Comment (mvdnes) [reply src/write.rs:140]: We could also just mark it as text for ASCII files. The appnote is a bit vague here:
```
       4.4.14.1 The lowest bit of this field indicates, if set, 
       that the file is apparently an ASCII or text file.  If not
       set, that the file apparently contains binary data.
```

**Comment by mvdnes** [comment]
I am fine leaving this out for now. I do not know of any use cases which actually read this flag, except for `zipinfo`. This was a quick addition to test if this fixes the epub problem, but it did not seem to have any effect.

**Comment by Plecra** [comment]
Thanks for bringing it up! We'll have to figure out a good solution in the end, but I don't think this is it - there are too many downsides. I'll leave the review to @rylev

**Comment by rylev** [comment]
@mvdnes @Plecra do we know what the use cases are where utf-8 (or even just ascii) data should be marked as a binary file?

**Comment by Plecra** [comment]
For now, we don't have any examples of code that would check the flag, no.

I'm fairly sure there'll be *some* zip-derived format that enforces the binary flag for, say, images, and images that technically contain valid unicode.

Also also, this will create false negatives too: "text" doesn't necessarily mean utf8.

**Comment by rylev** [comment]
What seems most reasonable to me in the long run: offer a way for the user to decide how files are classified with helpers for such things as declaring all valid utf8 as text. The default would be ASCII being text and everything else being marked binary.

**Comment by Plecra** [comment]
I'm moving this to issues as there haven't been any requests for the feature, and we're still not sure how best to implement it.

**Pull request closed without merging**